src/hotspot/share/opto/memnode.cpp

```diff
@@ -2770,9 +2770,461 @@ Node *StoreNode::Ideal(PhaseGVN *phase, bool can_reshape) {
+  // Marking the store mismatched is sufficient to prevent reordering, since array stores
+  // are all on the same slice. Hence, we need no barriers.
+  new_store->set_mismatched_access();
+
+  // Constants above may now also be be packed -> put candidate on worklist
+  phase->is_IterGVN()->_worklist.push(new_mem);
+
+#ifdef ASSERT
+  if (TraceMergeStores) {
+    stringStream ss;
+    ss.print_cr("[TraceMergeStores]: Replace");
+    for (int i = pow2size - 1; i >= 0; i--) {
+      merge_list.at(i)->dump("\n", false, &ss);
+    }
+    ss.print_cr("[TraceMergeStores]: with");
+    new_store->dump("\n", false, &ss);
+    tty->print("%s", ss.as_string());
+  }
+#endif
+
+  return new_store;
+}
+
+StoreNode* StoreNode::can_merge_primitive_array_store_with_use(PhaseGVN* phase, bool check_def) {
+  int opc = Opcode();
+  assert(opc == Op_StoreB || opc == Op_StoreC || opc == Op_StoreI, "precondition");
+
+  // Uses should be:
+  // 1) the other StoreNode
+  // 2) optionally a MergeMem from the uncommon trap
+  if (outcnt() > 2) {
+    return nullptr;
+  }
+
+  StoreNode* use_store = nullptr;
+  MergeMemNode* merge_mem = nullptr;
+  for (DUIterator_Fast imax, i = fast_outs(imax); i < imax; i++) {
+    Node* use = fast_out(i);
+    if (use->Opcode() == opc && use_store == nullptr) {
+      use_store = use->as_Store();
+    } else if (use->is_MergeMem() && merge_mem == nullptr) {
+      merge_mem = use->as_MergeMem();
+    } else {
+      return nullptr;
+    }
+  }
+  if (use_store == nullptr) {
+    return nullptr;
+  }
+  if (merge_mem != nullptr) {
+    // Check that merge_mem only leads to the uncommon trap between
+    // the two stores.
+    if (merge_mem->outcnt() != 1) {
+      return nullptr;
+    }
+    Node* ctrl_s1 = use_store->in(MemNode::Control);
+    Node* ctrl_s2 = this->in(MemNode::Control);
+    if (!ctrl_s1->is_IfProj() ||
+        !ctrl_s1->in(0)->is_RangeCheck() ||
+        ctrl_s1->in(0)->outcnt() != 2) {
+      return nullptr;
+    }
+    ProjNode* other_proj = ctrl_s1->as_IfProj()->other_if_proj();
+    Node* trap = other_proj->is_uncommon_trap_proj(Deoptimization::Reason_range_check);
+    if (trap != merge_mem->unique_out() ||
+        ctrl_s1->in(0)->in(0) != ctrl_s2) {
+      return nullptr;
+    }
+  }
+
+  // Having checked "def -> use", we now check "use -> def".
+  if (check_def) {
+    StoreNode* use_def = use_store->can_merge_primitive_array_store_with_def(phase, false);
+    if (use_def == nullptr) {
+      return nullptr;
+    }
+    assert(use_def == this, "def of use is this");
+  }
+
+  return use_store;
+}
+
+// Class to parse array pointers of the form:
+// pointer = base + constant_offset + (int_offset << int_offset_shift) + sum(other_offsets)
+//
+// The goal is to check if two such ArrayPointers are adjacent for a load or store.
+//
+// Note: we accumulate all constant offsets into constant_offset, even the int constant behind
```

Comment:
Is this really needed? For the patterns of interest, aren't the constant pushed down the chain of `AddP` nodes so the address is `(AddP base (AddP ...) constant)`?

Reply:
No, they are not pushed down.
Consider the access on an int array:
`a[invar + 1]` -> `adr = base + ARRAY_INT_BASE_OFFSET + 4 * ConvI2L(invar + 1)`
We cannot just push the constant `1` out of the `ConvI2L`, after all `invar + 1` could overflow in the int domain ;)

Reply:
That's not quite right, I think. For instance, in this method:
```
   private static int test(int[] array, int i) {
        return array[i + 1];
    }
```
the final IR will have the  `(AddP base (AddP ...) constant)` because `ConvI2LNode::Ideal` does more than checking for overflow. The actual transformation to that final shape must be delayed until after the CastII nodes are removed though. Why that's the case is puzzling actually because `CastIINode::Ideal()` has logic to push the AddI thru the `CastII` but it's disabled for range check `CastII` nodes. I noticed this while working on 8324517. My recollection was that `ConvI2LNode::Ideal` would push thru both the `CastII` and `ConvI2L` in one go so  I wonder if it got broken at some point.

Reply:
Thanks for info, I'll look into this :)

Reply:
Ah, I see what you are saying. The `AddI` can be pushed through the `ConvI2L`, but only because we know that the types are constrained. The types are constrained because of the `CastII` after the `RangeCheck`.